beacon_node/beacon_chain/src/shuffling_cache.rs

```diff
@@ -36,11 +36,38 @@ impl ShufflingCache {
         opt
     }
 
-    pub fn insert(&mut self, epoch: Epoch, root: Hash256, committee_cache: &CommitteeCache) {
-        let key = (epoch, root);
+    pub fn contains(&self, key: &ShufflingId) -> bool {
+        self.cache.contains(key)
+    }
 
+    pub fn insert(&mut self, key: ShufflingId, committee_cache: &CommitteeCache) {
         if !self.cache.contains(&key) {
             self.cache.put(key, committee_cache.clone());
         }
     }
 }
+
+/// Contains the shuffling IDs for a beacon block.
+pub struct BlockShufflingIds {
+    pub current: ShufflingId,
+    pub next: ShufflingId,
+}
+
+impl BlockShufflingIds {
+    /// Returns the shuffling ID for the given epoch.
+    ///
+    /// Returns `None` if `epoch` is prior to `self.current.shuffling_epoch`.
+    pub fn id_for_epoch(&self, epoch: Epoch) -> Option<ShufflingId> {
+        if epoch == self.current.shuffling_epoch {
+            Some(self.current.clone())
+        } else if epoch == self.next.shuffling_epoch {
+            Some(self.next.clone())
+        } else if epoch > self.next.shuffling_epoch {
+            let mut shuffling_id = self.next.clone();
+            shuffling_id.shuffling_epoch = epoch;
+            Some(shuffling_id)
+        } else {
```

Comment:
This doesn't seem safe to me... if we have the shuffling IDs for the current and next epochs, why can we assume we know the shuffling for future epochs?

Reply:
I kinda convinced myself that this was OK during our call, but I think it's still not quite right.

Our goal here is to return a shuffling ID that accurately describes the shuffling at a future epoch where no new blocks have been produced since the given head block. The problem is, the shuffling ID for the next epoch (from the perspective of the head), only includes RANDAO information up to the last slot of the previous epoch. The shuffling for subsequent epochs (>`next_epoch`) is affected by blocks _after_ that point, namely those in the current epoch, including the head block itself. Diagram for clarity:

![](https://i.imgur.com/oLjRtNN.jpg)

I think the correct shuffling ID to return for an `epoch` after `next_epoch` would be:

```rust
ShufflingId {
    shuffling_epoch: epoch,
    shuffling_decision_block: head_block_root,
}
```

Which would require incorporating the `head_block_root` into the `BlockShufflingIds` struct.

Reply:
Yep, you're totally right. Thank you!